[Style] Apply zoom to lengths within calc when evaluating.

sammygill opened the pull request:
#### 761924a0502321e047299221e9ef26096458dc8b
<pre>
[Style] Apply zoom to lengths within calc when evaluating.
<a href="https://bugs.webkit.org/show_bug.cgi?id=301219">https://bugs.webkit.org/show_bug.cgi?id=301219</a>
<a href="https://rdar.apple.com/problem/163141549">rdar://problem/163141549</a>

Reviewed by Sam Weinig.

Before we can start evaluting Preferred, Minimum, and Maximum Sizes at
use time with zoom we need to be able to support evaluating calc
expressions with zoom since any length values within the expression
should be affected by zoom. For example, something like
width: calc(50% + 100px) will need to have the used zoom multiplied into
the 100px.

We can do this in a similar manner to how we force callers to pass in a
ZoomFactor to various Length types when evaluating their values. The
Style Evaluation code will then call into the Calculation&apos;s evaluate API
with the passed in zoom factor. This requires us to plumb the ZoomFactor
through this codepath when evaluating different Calc nodes and using it
when applicable. The other half of this patch is then mostly applying
Style::ZoomNeeded { } at different call sites to mark which types still
need to become Unzoomed and have an actual ZoomFactor passed in.

* LayoutTests/fast/css/line-height-calc-zoomed-expected.html: Added.
* LayoutTests/fast/css/line-height-calc-zoomed.html: Added.
* Source/WebCore/layout/formattingContexts/inline/InlineLevelBox.h:
(WebCore::Layout::InlineLevelBox::preferredLineHeight const):
Here we can actually pass in the usedZoom since line-height is already
Unzoomed.

* Source/WebCore/rendering/RenderBox.cpp:
(WebCore::RenderBox::computePercentageLogicalHeightGeneric const):
Note here we separate out the percent and calc logic since we need to
now treat these differently. Percentages still do not have zoom applied
to them but calc expressions will now have the used zoom applied to
their lengths. This is done in a few different area of the code but I am
only going to note it here.

(WebCore::RenderBox::computeOutOfFlowPositionedLogicalWidthUsing const):
* Source/WebCore/rendering/RenderFlexibleBox.cpp:
(WebCore::RenderFlexibleBox::computeMainSizeFromAspectRatioUsing const):
* Source/WebCore/rendering/RenderGrid.cpp:
(WebCore::RenderGrid::computeAutoRepeatTracksCount const):
* Source/WebCore/rendering/RenderReplaced.cpp:
(WebCore::RenderReplaced::paint):
* Source/WebCore/rendering/style/RenderStyle.cpp:
(WebCore::RenderStyle::computeLineHeight const):
(WebCore::RenderStyle::computedStrokeWidth const):
* Source/WebCore/style/StyleBuilderCustom.h:
(WebCore::Style::BuilderCustom::applyValueFontSize):
* Source/WebCore/style/StyleExtractorCustom.h:
(WebCore::Style::PropertyExtractorAdaptor&lt;CSSPropertyLineHeight&gt;::computedValue const):

* Source/WebCore/style/calc/StyleCalculationTree+Evaluation.cpp:
(WebCore::Style::Calculation::evaluate):
* Source/WebCore/style/calc/StyleCalculationTree+Evaluation.h:
* Source/WebCore/style/calc/StyleCalculationValue.cpp:
(WebCore::Style::Calculation::Value::evaluate const):
* Source/WebCore/style/calc/StyleCalculationValue.h:
Plumbing for the zoom factor through all the different evaluate
implementations. This should only get applied to lengths so in many
cases we do not need to use it.

* Source/WebCore/style/values/primitives/StyleLengthWrapperData.h:
(WebCore::Style::LengthWrapperData::minimumValueForLengthWrapperDataWithLazyMaximum const):
(WebCore::Style::LengthWrapperData::valueForLengthWrapperDataWithLazyMaximum const):
* Source/WebCore/style/values/primitives/StylePrimitiveNumeric.h:
* Source/WebCore/style/values/primitives/StylePrimitiveNumericTypes+Evaluation.h:
When called on a type that is Unzoomed we will pass in the ZoomFactor
that they gave us, but in the other case we will just use a
Style::ZoomFactor { 1.0f, 1.0f, } so that the calc value is unaffected.
* Source/WebCore/style/values/scroll-snap/StyleScrollPadding.cpp:
(WebCore::Style::LayoutUnit&gt;::operator):
(WebCore::Style::float&gt;::operator):
* Source/WebCore/style/values/text-decoration/StyleTextUnderlineOffset.cpp:
(WebCore::Style::TextUnderlineOffset::resolve const):
* Source/WebCore/svg/SVGLengthContext.cpp:
(WebCore::SVGLengthContext::valueForSizeType):

Canonical link: <a href="https://commits.webkit.org/301968@main">https://commits.webkit.org/301968@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/2628c9ee487674a34d85029f3a023cc4422f93eb

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/127626 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/47274 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/38442 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/134897 "Built successfully") | [  ~~🛠 win~~](https://ews-build.webkit.org/#/builders/59/builds/79183 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 ios-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/aed576a0-4dfd-47cb-9743-2130abaafee1) 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/129498 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/47897 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/55804 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/97155 "Passed tests") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/59/builds/79183 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 mac-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/45ced037-295e-406d-bec4-d96dc4884da4) 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/130574 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/38308 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/114327 "Passed tests") | [❌ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/77636 "Found 1 new API test failure: WPE/TestWebKitWebXR:/webkit/WebKitWebXR/permission-request (failure)") | | [✅ 🛠 vision-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/8c03458c-7f5e-4ad8-85c7-7ac630e912a8) 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/37115 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/32416 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/78256 "Built successfully") | | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/108176 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/32867 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/137379 "Built successfully") | | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/54285 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/41842 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/105675 "Passed tests") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/54796 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/110682 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/105326 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/26860 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/50845 "Passed tests") | [  ~~🧪 mac-intel-wk2~~](https://ews-build.webkit.org/#/builders/137/builds/29279 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/51879 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/54222 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/60398 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/53456 "Built successfully") | | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/56913 "Built successfully") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/55215 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

weinig commented on Source/WebCore/style/values/primitives/StyleLengthWrapperData.h:
```diff
@@ -284,7 +284,7 @@ ReturnType LengthWrapperData::minimumValueForLengthWrapperDataWithLazyMaximum(Le
     case LengthWrapperDataEvaluationKind::Calculation:
         ASSERT(m_kind == LengthWrapperDataKind::Calculation);
-        return ReturnType(nonNanCalculatedValue(lazyMaximumValueFunctor()));
+        return ReturnType(nonNanCalculatedValue(lazyMaximumValueFunctor(), ZoomFactor { 1.0f, 1.0f }));
```
I would like us to avoid doing this anywhere if we can.

The idiom established is to overload the function that needs with one that takes Style::ZoomNeeded. 

So in this case, there should be a `float nonNanCalculatedValue(float maxValue, Style::ZoomNeeded)` overload and you would pass `Style::ZoomNeeded { }` here instead of the `1.0f`.

This helps find places that still need to be updated in a consistent way.

sammygill replied:
Makes sense to me, changed! Along with the other places that now have an overloaded variant.

weinig commented on Source/WebCore/style/values/primitives/StylePrimitiveNumericTypes+Evaluation.h:
```diff
@@ -94,16 +94,24 @@ template<NonCompositeNumeric StyleType, typename Result> struct Evaluation<Style
+        requires (Calculation::range.zoomOptions == CSS::RangeZoomOptions::Default)
     {
-        return evaluate<Result>(calculation.protectedCalculation(), referenceLength, std::forward<Rest>(rest)...);
+        return evaluate<Result>(calculation.protectedCalculation(), referenceLength, ZoomFactor { 1.0f, 1.0f }, std::forward<Rest>(rest)...);
```
Same here, please pass along the ZoomNeeded token.

weinig commented on Source/WebCore/style/values/text-decoration/StyleTextUnderlineOffset.cpp:
```diff
@@ -42,6 +42,9 @@ float TextUnderlineOffset::resolve(const RenderStyle& style, float autoValue) co
+        [&](const Calc& calc) -> float {
+            return Style::evaluate<float>(calc, style.computedFontSize(), style.usedZoomForLength());
+        },
         [&](const auto& percentage) -> float {
```
Mind making this `Percentage` instead of `auto`. Mostly try to avoid `auto` in these, as you lose the compile errors when new alternatives are added.

sammygill replied:
Yup, done

weinig commented on Source/WebCore/style/calc/StyleCalculationTree+Evaluation.cpp:
```diff
@@ -57,102 +58,106 @@ template<typename Op, typename... Args> static double executeMathOperation(Args&
 }
 
-double evaluate(const Tree& tree, double percentResolutionLength)
+double evaluate(const Tree& tree, double percentResolutionLength, const ZoomFactor& zoomFactor)
```
This function should be overloaded with a ZoomNeeded version, but the rest of the functions in this file don't, that would be excessive.

sammygill replied:
Done! Since this is a bottleneck function and we want to avoid modifying all of the other functions I guess this overload might need to pass in a `Style::ZoomFactor { 1.0f, 1.0f }` unless you think there is a better way without touching the rest?

weinig replied:
Nope, you have to do it there. You’re right.

weinig commented on Source/WebCore/style/calc/StyleCalculationValue.h:
```diff
@@ -51,7 +54,7 @@ class Value : public RefCounted<Value> {
     WEBCORE_EXPORT ~Value();
 
-    double evaluate(double percentResolutionLength) const;
+    double evaluate(double percentResolutionLength, const ZoomFactor& usedZoom) const;
```
This should be overloaded with a ZoomNeeded version.

weinig commented on Source/WebCore/style/values/primitives/StyleLengthWrapperData.h:
```diff
@@ -96,7 +96,7 @@ struct LengthWrapperData {
 
 private:
-    WEBCORE_EXPORT float nonNanCalculatedValue(float maxValue) const;
+    WEBCORE_EXPORT float nonNanCalculatedValue(float maxValue, const ZoomFactor& usedZoom) const;
```
This should be overload with a ZoomNeeded version.

weinig commented on Source/WebCore/rendering/RenderReplaced.cpp:
```diff
@@ -1174,6 +1181,9 @@ LayoutUnit RenderReplaced::computeReplacedLogicalHeightUsingGeneric(const SizeTy
 
     auto percentageOrCalculated = [&](const auto& logicalHeight) {
+        using T = std::decay_t<decltype(logicalHeight)>;
+        static_assert(Style::IsPercentage<T> || Style::IsCalc<T>);
```
If you wanted to be fancy, you could make this a constraint instead.

```
    auto percentageOrCalculated = [&](Style::IsPercentageOrCalc auto const& logicalHeight) {
```

brentfulgham commented on Source/WebCore/style/StyleBuilderCustom.h:
```diff
@@ -1269,7 +1269,7 @@ inline void BuilderCustom::applyValueFontSize(BuilderState& builderState, CSSVal
             size = (primitiveValue->resolveAsPercentage<float>(conversionData) * parentSize) / 100.0f;
         else if (primitiveValue->isCalculatedPercentageWithLength())
-            size = primitiveValue->cssCalcValue()->createCalculationValue(conversionData, CSSCalcSymbolTable { })->evaluate(parentSize);
+            size = primitiveValue->cssCalcValue()->createCalculationValue(conversionData, CSSCalcSymbolTable { })->evaluate(parentSize, Style::ZoomFactor { conversionData.zoom(), 1.0f });
```
Not a lot of platforms where the display is 1.0x!

weinig replied:
I think you probably want ZoomNeeded { } here.

weinig commented on Source/WebCore/rendering/RenderReplaced.cpp:
```diff
@@ -1056,11 +1056,15 @@ LayoutUnit RenderReplaced::computeReplacedLogicalWidthUsing(const SizeType& logi
     };
 
-    auto percentageOrCalc = [&](const auto& logicalWidth) {
+    auto percentageOrCalc =  [&](Style::IsPercentageOrCalc auto const& logicalWidth) {
```
Unintentional space added before `[&]`.

weinig commented on Source/WebCore/style/StyleExtractorCustom.h:
```diff
@@ -283,7 +283,7 @@ template<> struct PropertyExtractorAdaptor<CSSPropertyLineHeight> {
             },
             [&](const LineHeight::Calc& calc) {
-                return functor(Length<CSS::Nonnegative> { evaluate<float>(calc, 0.0f) });
+                return functor(Length<CSS::Nonnegative> { evaluate<float>(calc, 0.0f, Style::ZoomFactor { 1.0f, 1.0f }) });
```
Hm, this one is weird. I actually can't imagine why we are evaluating calc() here in the extractor, but LineHeight is very unusual. Can you add a comment here stating that passing 1.0f here is to get the unzoomed value needed for computed value and that we should come back and re-evaluate how to do this in the future?

weinig commented on Source/WebCore/style/values/images/StyleGradient.cpp:
```diff
@@ -155,7 +155,7 @@ static std::optional<float> resolveColorStopPosition(const GradientAngularColorS
         },
         [&](const typename AnglePercentage<>::Calc& calc) -> std::optional<float> {
-            return calc.protectedCalculation()->evaluate(100) / 100.0;
+            return Style::evaluate<float>(calc, 100, Style::ZoomNeeded { });
```
I don't think a AnglePercentage will ever have a length at this point (it will only be degrees or percentages in the calc), but I think this is fine. We will need to make it more clear in StyleCalculation that only sometimes does it make sense to multiple the dimension by zoom. I will need to think more about how to do that cleanly.

weinig commented on Source/WebCore/style/values/primitives/StyleLengthWrapperData.cpp:
```diff
@@ -81,10 +81,20 @@ auto LengthWrapperData::ipcData() const -> IPCData
+    return result;
+}
+
+
```
Extra newline.

sammygill committed:
[Style] Apply zoom to lengths within calc when evaluating.
https://bugs.webkit.org/show_bug.cgi?id=301219
rdar://problem/163141549

Reviewed by Sam Weinig.

Before we can start evaluting Preferred, Minimum, and Maximum Sizes at
use time with zoom we need to be able to support evaluating calc
expressions with zoom since any length values within the expression
should be affected by zoom. For example, something like
width: calc(50% + 100px) will need to have the used zoom multiplied into
the 100px.

We can do this in a similar manner to how we force callers to pass in a
ZoomFactor to various Length types when evaluating their values. The
Style Evaluation code will then call into the Calculation's evaluate API
with the passed in zoom factor. This requires us to plumb the ZoomFactor
through this codepath when evaluating different Calc nodes and using it
when applicable. The other half of this patch is then mostly applying
Style::ZoomNeeded { } at different call sites to mark which types still
need to become Unzoomed and have an actual ZoomFactor passed in.

* LayoutTests/fast/css/line-height-calc-zoomed-expected.html: Added.
* LayoutTests/fast/css/line-height-calc-zoomed.html: Added.
* Source/WebCore/layout/formattingContexts/inline/InlineLevelBox.h:
(WebCore::Layout::InlineLevelBox::preferredLineHeight const):
Here we can actually pass in the usedZoom since line-height is already
Unzoomed.

* Source/WebCore/rendering/RenderBox.cpp:
(WebCore::RenderBox::computePercentageLogicalHeightGeneric const):
Note here we separate out the percent and calc logic since we need to
now treat these differently. Percentages still do not have zoom applied
to them but calc expressions will now have the used zoom applied to
their lengths. This is done in a few different area of the code but I am
only going to note it here.

(WebCore::RenderBox::computeOutOfFlowPositionedLogicalWidthUsing const):
* Source/WebCore/rendering/RenderFlexibleBox.cpp:
(WebCore::RenderFlexibleBox::computeMainSizeFromAspectRatioUsing const):
* Source/WebCore/rendering/RenderGrid.cpp:
(WebCore::RenderGrid::computeAutoRepeatTracksCount const):
* Source/WebCore/rendering/RenderReplaced.cpp:
(WebCore::RenderReplaced::paint):
* Source/WebCore/rendering/style/RenderStyle.cpp:
(WebCore::RenderStyle::computeLineHeight const):
(WebCore::RenderStyle::computedStrokeWidth const):
* Source/WebCore/style/StyleBuilderCustom.h:
(WebCore::Style::BuilderCustom::applyValueFontSize):
* Source/WebCore/style/StyleExtractorCustom.h:
(WebCore::Style::PropertyExtractorAdaptor<CSSPropertyLineHeight>::computedValue const):

* Source/WebCore/style/calc/StyleCalculationTree+Evaluation.cpp:
(WebCore::Style::Calculation::evaluate):
* Source/WebCore/style/calc/StyleCalculationTree+Evaluation.h:
* Source/WebCore/style/calc/StyleCalculationValue.cpp:
(WebCore::Style::Calculation::Value::evaluate const):
* Source/WebCore/style/calc/StyleCalculationValue.h:
Plumbing for the zoom factor through all the different evaluate
implementations. This should only get applied to lengths so in many
cases we do not need to use it.

* Source/WebCore/style/values/primitives/StyleLengthWrapperData.h:
(WebCore::Style::LengthWrapperData::minimumValueForLengthWrapperDataWithLazyMaximum const):
(WebCore::Style::LengthWrapperData::valueForLengthWrapperDataWithLazyMaximum const):
* Source/WebCore/style/values/primitives/StylePrimitiveNumeric.h:
* Source/WebCore/style/values/primitives/StylePrimitiveNumericTypes+Evaluation.h:
When called on a type that is Unzoomed we will pass in the ZoomFactor
that they gave us, but in the other case we will just use a
Style::ZoomFactor { 1.0f, 1.0f, } so that the calc value is unaffected.
* Source/WebCore/style/values/scroll-snap/StyleScrollPadding.cpp:
(WebCore::Style::LayoutUnit>::operator):
(WebCore::Style::float>::operator):
* Source/WebCore/style/values/text-decoration/StyleTextUnderlineOffset.cpp:
(WebCore::Style::TextUnderlineOffset::resolve const):
* Source/WebCore/svg/SVGLengthContext.cpp:
(WebCore::SVGLengthContext::valueForSizeType):

Canonical link: https://commits.webkit.org/301968@main
